zoom_rooms/indico_zoom_rooms/plugin.py

```diff
@@ -0,0 +1,271 @@
+            f(obj, *args, **kwargs)
+
+    return _wrapper
+
+
+@only_zoom_rooms
+def _delete_zoom_association(old_link: SessionBlock | Contribution | Event, vc_room: VCRoom):
+    """Handle deleted VC room associations (given the corresponding object)."""
+    if zr_id := get_zoom_room_id(old_link.room):
+        ZoomRoomsQueueEntry.record(ZoomRoomsAction.delete, zr_id, obj=old_link, vc_room=vc_room)
+
+
+@only_zoom_rooms
+def _handle_link_object_created(assoc: VCRoomEventAssociation):
+    """Handle the creation of a new object (given the corresponding association)."""
+    if assoc.vc_room.type == 'zoom':
+        if zr_id := get_zoom_room_id(assoc.link_object.room):
+            ZoomRoomsQueueEntry.record(ZoomRoomsAction.create, zr_id, assoc=assoc)
+
+
+@only_zoom_rooms
+def _handle_link_object_dt_change(
+    obj: SessionBlock | Contribution | Event,
+    changes: dict[str, t.Any],
+):
+    """Handle the changes of date/time of an object."""
+    if zr_id := get_zoom_room_id(obj.room):
+        # do not handle sub-objects, as the corresponding signals are called directly
+        for assoc in obj.vc_room_associations:
+            if assoc.vc_room.type == 'zoom':
+                args = OperationArgs()
+                if 'start_dt' in changes:
+                    args['start_dt'] = int(changes['start_dt'][1].timestamp())
+                if 'end_dt' in changes:
+                    args['end_dt'] = int(changes['end_dt'][1].timestamp())
+
+                ZoomRoomsQueueEntry.record(ZoomRoomsAction.update, zr_id, assoc=assoc, args=args)
+
+
+def _handle_contribution_move(obj: Contribution, original_block: SessionBlock, new_block: SessionBlock):
+    """Handle contributions moved between blocks."""
+    # block is None -> attached to the event
+    if (orig_zr_id := get_zoom_room_id(original_block.room if original_block else obj.event.room)) != (
+        new_zr_id := get_zoom_room_id(new_block.room if new_block else obj.event.room)
+    ):
+        _handle_linked_obj_location_change(obj, orig_zr_id, new_zr_id)
+
+
+def _handle_linked_obj_location_change(
+    obj: SessionBlock | Contribution | Event,
+    old_zr_id: str | None,
+    new_zr_id: str | None,
+):
+    """Handle changes of location in an object (given the old/new ZoomRoom ID)."""
+    match old_zr_id, new_zr_id:
+        case None, None:
+            # this doesn't concern us as there are no zoom rooms involved
+            return
+        case None, zr_id:
+            make_op = lambda assoc: ZoomRoomsQueueEntry.record(ZoomRoomsAction.create, t.cast(str, zr_id), assoc=assoc)
+        case zr_id, None:
+            make_op = lambda assoc: ZoomRoomsQueueEntry.record(ZoomRoomsAction.delete, t.cast(str, zr_id), assoc=assoc)
+        case old_zr_id, new_zr_id:
+            make_op = lambda assoc: ZoomRoomsQueueEntry.record(
+                ZoomRoomsAction.move, t.cast(str, old_zr_id), assoc=assoc, args={'new_zr_id': t.cast(str, new_zr_id)}
+            )
```

Comment:
```suggestion
        case old_id, new_id:
            make_op = lambda assoc: ZoomRoomsQueueEntry.record(
                ZoomRoomsAction.move, t.cast(str, old_id), assoc=assoc, args={'new_zr_id': t.cast(str, new_id)}
            )
```

I don't feel really strongly about it, but it took me a moment to notice htat it's just the catchall case where it re-assigns the same values to the same names...

Reply:
Not sure what you mean here. Is it just the name?

Reply:
yeah, since you're basically doing the equivalent of `old_zr_id = old_zr_id` for that block. but like i said, no strong feeling about it, not sure what's actually cleaner

Reply:
Interesting, to me having the same name actually helps me realize it's a 1-to-1 mapping. Also no strong opinion. An alternative could be:

```python
       case _:
            make_op = lambda assoc: ZoomRoomsQueueEntry.record(
                ZoomRoomsAction.move, t.cast(str, old_zr_id), assoc=assoc, args={'new_zr_id': t.cast(str, new_zr_id)}
            )
```
But TBH it has its shortcomings too...